### PR TITLE
[WIP] Use PEP 517 Build Isolation

### PR DESCRIPTION
**Pull request by Cadair**
## Description

This is following on from a discussion in #4261 

This works for me currently when installing from a checkout in a completely clean venv. What I am unclear on at this point is how the C files are generated for the sdist, I am proposing that we remove that behaviour and have people generate the C files during build (with a pinned known version of Cython).

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[build-system]
+requires = ["setuptools", "wheel", "numpy==1.14.5", "cython==0.29.0"]
```
Comment (Cadair) [inline pyproject.toml]: I have pinned numpy 1.14.5 as it's the first numpy version with py37 wheels on pypi. I have no idea what a good Cython pin is, there is no reason why we couldn't go very new here if we needed to.

File: setup.py [hunk 2]
```diff
@@ -163,39 +166,16 @@ def configuration(parent_package='', top_path=None):
...
-        else:
-            print('To install scikit-image from source, you will need numpy.\n' +
-                  'Install numpy with pip:\n' +
-                  'pip install numpy\n'
-                  'Or use your operating system package manager. For more\n' +
-                  'details, see https://scikit-image.org/docs/stable/install.html')
-            sys.exit(1)
+    from numpy.distutils.core import setup
+    extra = {'configuration': configuration}
+    # Do not try and upgrade larger dependencies
```
Comment (Cadair) [inline setup.py]: This feels very weird to me, but I have left it alone.

File: pyproject.toml (hunk 1 above)
Comment (hmaarrfk) [reply pyproject.toml]: This works for python 3.8?

Comment (grlee77) [reply pyproject.toml]: I think the strategy is generally to choose the minimum NumPy that is both compatible with scikit-image and for which wheels exist. That is why SciPy has all of the following cases:

https://github.com/scipy/scipy/blob/6644a54ff862e4d90d7f9bc43ece9b0b1f681861/pyproject.toml#L5-L13

Comment (grlee77) [reply pyproject.toml]: Also, I think Cython >= 0.29.13 is the minimum compatible with Python 3.8.

Comment (Cadair) [reply pyproject.toml]: yes:

```
$ python --version
Python 3.8.0
$ pip freeze
pkg-resources==0.0.0
$ pip install .
```
numpy is now availible as a wheel for 3.8 but Cython 0.29.0 isn't, so I assume it has compiled it from source.

Comment (Cadair) [reply pyproject.toml]: I know that this build isolation used to rely on wheels existing, but it seems from this experiment that that is no longer the case. I wonder if I can find a pip reference for that somewhere.

**Pull request closed without merging**